File: dsa-problems/GameTheory/solutions/python/GMT-001-pile-split-choice.py
def pile_split_game(n: int) -> str:
    if n == 0:
        return "Second"
    
    g = [0] * (n + 1)
    
    for i in range(3, n + 1):
        reachable = set()
        # Split into j and i-j.
        # j goes from 1 up to (i-1)//2
        for j in range(1, (i - 1) // 2 + 1):
            reachable.add(g[j] ^ g[i - j])
        
        mex = 0
        while mex in reachable:
            mex += 1
        g[i] = mex
        
    return "First" if g[n] > 0 else "Second"

def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    n = int(data[0])
    print(pile_split_game(n))

if __name__ == "__main__":
    main()

File: dsa-problems/GameTheory/solutions/java/GMT-001-pile-split-choice.java
import java.util.*;

class Solution {
    public String pileSplitGame(int n) {
        if (n == 0) return "Second";
        int[] g = new int[n + 1];
        
        for (int i = 3; i <= n; i++) {
            Set<Integer> reachable = new HashSet<>();
            // Split into j and i-j. j starts at 1.
            // Condition: j != i-j => 2*j != i.
            // Also j < i-j to avoid duplicates => 2*j < i.
            for (int j = 1; 2 * j < i; j++) {
                reachable.add(g[j] ^ g[i - j]);
            }
            
            // Calculate mex
            int mex = 0;
            while (reachable.contains(mex)) {
                mex++;
            }
            g[i] = mex;
        }
        
        return g[n] > 0 ? "First" : "Second";
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            Solution solution = new Solution();
            System.out.println(solution.pileSplitGame(n));
        }
        sc.close();
    }
}
